### DISCUSSION TITLE
Native AOT and Avalonia

### DISCUSSION THREAD

**Discussion by totiev** [AvaloniaUI/Avalonia#8130] [Ideas]
It is possible use Avalonia with Native AOT ?

As far as I know current Native AOT does not support WPF and no nearest plans to support it. It seems to me it will be killer feature for Avalonia to support Native AOT.

**Comment by hez2010** [comment]
You can already use Avalonia with NativeAOT.

**Reply by totiev** [reply]
It really means that now I can develop GUI Avalonia applications in native machine codes?

**Reply by maxkatz6** [reply]
Yes. With current stable version there are minor problems, but basically this repo is a great example how to solve them https://github.com/teobugslayer/AvaloniaCoreRTDemo
With master branch (and future 11.0) it works without any special configs.

**Reply by totiev** [reply]
Thank you very much for this demo.

**Reply by uno-studio** [reply]
> Yes. With current stable version there are minor problems, but basically this repo is a great example how to solve them https://github.com/teobugslayer/AvaloniaCoreRTDemo With master branch (and future 11.0) it works without any special configs.


It seems that Binding can't work with Native AOT. after updating Avalonia app to .Net 7 and Native AOT, Binding bindings can't work, and I get warning:
`Trim analysis warning IL2026: AOTProject.Views.MenuView.!XamlIlPopulate(IServiceProvider,MenuView): Using member 'Avalonia.Mark
up.Xaml.MarkupExtensions.ReflectionBindingExtension.ReflectionBindingExtension(String)' which has 'RequiresUnreferenced
CodeAttribute' can break functionality when trimming application code. BindingExpression and ReflectionBinding heavily
use reflection. Consider using CompiledBindings instead.`
I still compile successfully but the application cannot display the Binding data. I still haven't found a workaround, can someone help?

**Reply by hez2010** [reply]
You need to use CompiledBindings.

**Reply by uno-studio** [reply]
> You need to use CompiledBindings.

In my application that uses converter and localizer, and cannot use all CompiledBindings

**Reply by hez2010** [reply]
I don't think there is any compatibility issue with CompiledBindings and converters. To use CompiledBindings you just need to specify the type explicitly using x:Type.

**Reply by uno-studio** [reply]
> I don't think there is any compatibility issue with CompiledBindings and converters. To use CompiledBindings you just need to specify the type explicitly using x:Type.

localizer use this code to binding
```
public class LocalizeExtension : MarkupExtension
{
    public LocalizeExtension(string key)
    {
        Key = key;
    }

    public string Key { get; set; }

    public string Context { get; set; }

    public override object ProvideValue(IServiceProvider serviceProvider)
    {
        var keyToUse = Key;
        if (!string.IsNullOrWhiteSpace(Context))
            keyToUse = $"{Context}/{Key}";

        var binding = new ReflectionBindingExtension($"[{keyToUse}]")
        {
            Mode = BindingMode.OneWay,
            Source = Localizer.Instance
        };

        return binding.ProvideValue(serviceProvider);
    }
}
````

**Reply by hez2010** [reply]
Then you can use TrimmerRootDescriptor to workaround.

**Reply by danielmayost** [reply]
I also encountered this, there is two more options:
1. Use CompiledBinding instead ReflectionBinding in your MarkupExtension (it's going to be hard, this can help: https://github.com/AvaloniaUI/Avalonia/discussions/10077#discussioncomment-4813935).
2. Expose the Localizer reference in the ViewModelBase so you can access it from any View in a direct binding, that's what I did too.

**Reply by Cubody** [reply]
Is it possible to publish it as single exe? It always need Skia and Harfbuzz libraries.

**Reply by hez2010** [reply]
If you have the static library of skia and harfbuss, you can statically link those libraries into the binary.

**Reply by Cubody** [reply]
Can you tell me how to do this, please? Or provide a link to useful documentation.

**Reply by maxkatz6** [reply]
Single file with Native AOT isn't a supported scenario by .NET runtime itself. Maybe in the future they will support it. You can ask at https://github.com/dotnet/runtime. Avalonia is not related here.

As mentioned, you can static link libraries, but you would need to build themselves by yourself - https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/using-nativeaot/interop.md#linking (and some details here https://github.com/AvaloniaUI/Avalonia/issues/9503).

Also, you can find some tools that wrap executable and files into some kind of self extracting archive.

**Reply by Cubody** [reply]
Thanks for the answer!

**Reply by cyraid** [reply]
> Single file with Native AOT isn't a supported scenario by .NET runtime itself. Maybe in the future they will support it. You can ask at https://github.com/dotnet/runtime. Avalonia is not related here.
> 
> As mentioned, you can static link libraries, but you would need to build themselves by yourself - https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/using-nativeaot/interop.md#linking (and some details here #9503).
> 
> Also, you can find some tools that wrap executable and files into some kind of self extracting archive.

I've built Native AOT single file executables.

@Cubody 
As for Skia, you'd have to get it's backing library as well, and add some `<DirectPInvoke Include="lib" />` and `<NativeLibrary Include="skialibhere.lib" />` and probably user32.lib, gdi32.lib, shell32.lib you can see when you build what unresolved symbols there are. I plan to do this eventually too for my project just haven't gotten around to it yet.

When you build SkiaSharp, you might be able to grab the .lib files from the various harfbuzz, skia, and skiasharp glue library. If not I'm sure there's some compiling flags you can set somewhere to statically build.

For windows you might need to throw in `<LinkerArg Include="/NODEFAULTLIB:LIBCMT" />` if it gives you the default library error.

**Reply by PRNDA** [reply]
How do you built Native AOT single file executables, could you please explain some more?

**Reply by DeafMan1983** [reply]
Both libraries should compile into static files *.lib for Windows and *.a files for macOS/Linux and we would like to embed into native single executable without loading dynamic libraries.

**Reply by danielmayost** [reply]
> How do you built Native AOT single file executables, could you please explain some more?

the simplest (for windows), download the lib files from here: https://github.com/AvaloniaUI/Avalonia/issues/9503#issuecomment-1435511465, and then add these lines to csproj file:
```
 <ItemGroup>
   <DirectPInvoke Include="libHarfBuzzSharp" />
   <NativeLibrary Include="native\libHarfBuzzSharp.lib" Condition="$(RuntimeIdentifier.StartsWith('win'))" />
   <DirectPInvoke Include="libSkiaSharp" />
   <NativeLibrary Include="native\libSkiaSharp.lib" Condition="$(RuntimeIdentifier.StartsWith('win'))" />
 </ItemGroup>
```

To get rid of `av_libglesv2.dll`, you will probably need to use Wgl. (Add `UseWgl = true` in your `Win32PlatformOptions` in `AppBuilder`), i couldn't compile ANGLE statically.

**Reply by DeafMan1983** [reply]
Ps why do you not use wgl from gdi32.lib or opengl32.lib? And for Vulkan should use vulkan.lib

**Reply by danielmayost** [reply]
these files exists every windows system, why you need them statically? the goal is to publish single file without any external dlls.

**Reply by hez2010** [reply]
> why do you not use wgl

To be short, wgl is somewhat deprecated on Windows and it has too many bugs and missing features.

**Reply by DeafMan1983** [reply]
Many bugs? Why not wrapper with glew 2.2x

**Reply by hez2010** [reply]
> Why not wrapper with glew 2.2x

By using the native OpenGL implementation, you will experience endless bugs from different graphics drivers and different vendors. So we instead use ANGLE, which translates OpenGL calls to DirectX on Windows (and to Metal on macOS) to deliver a more stable, polished and consistent experience on all platforms.

**Reply by Gillibald** [reply]
Like Chrome does...

**Reply by DeafMan1983** [reply]
Why not use emscripten with EGL 1.5 that's OpenGL ES for chrome or Firefox

**Reply by Gillibald** [reply]
You are confusing parts here. Emscripten exposes such API to the consumer. ANGLE is used to drive the rendering on the OS level.

Emscripten -> Sandbox
ANGLE -> OS

**Reply by Cubody** [reply]
Hey, so there is no way to use single file with ANGLE? Or it is possible to build .lib?

**Reply by maxkatz6** [reply]
The biggest problem with angle, is to get this lib file. If there is a guide on how to build all required lib files for statically linked Angle, it can be done and shipped with the nuget package.

**Reply by NCLnclNCL** [reply]
Yes

**Reply by hez2010** [reply]
> it can be done and shipped with the nuget package

The static lib of Angle yields a 1.3G .lib file. I don't think shipping such a huge file on the nuget is a good thing... 
Instead how about authoring a msbuild target to build Angle locally?

**Reply by Cubody** [reply]
Can't see any problem about size

**Comment by peaceshi** [comment]
> The biggest problem with angle, is to get this lib file. If there is a guide on how to build all required lib files for statically linked Angle, it can be done and shipped with the nuget package.

A suitable way to build ANGLE is using vcpkg.
`vcpkg install angle:x64-windows-static`

I try to build it, and everything works well.

Because it is easily to build with vcpkg, I doesn't upload the lib here.

![image](https://github.com/AvaloniaUI/Avalonia/assets/10380828/3bfe2372-89d0-470b-b21f-6a656c84bf59)

**Reply by peaceshi** [reply]
I have created two github actions to build these static libs.

libSkiaSharp and libHarfBuzzSharp
https://github.com/2ndlab/SkiaSharp.Static

ANGLE static
https://github.com/2ndlab/ANGLE.Static

**Reply by DeafMan1983** [reply]
Does it work on Linux/BSD?

**Reply by peaceshi** [reply]
definitely impossible. It works on windows only.
It is strongly recommended **NOT** to be used by those unfamiliar with native programming.

**Reply by hez2010** [reply]
> definitely impossible

No. It's possible. But to use it on Linux you need to rebuild and target muslc instead of glibc.

**Reply by DeafMan1983** [reply]
Thanks if you use Alpine or Void-Linux than you need to compile with `musl-libc` and OpenSuse, Ubuntu, RedHat or other with GNU `glibc`

**Reply by hez2010** [reply]
glibc is not compatible with static linking and can cause crashing at runtime.

**Comment by sgf** [comment]
In fact, I don't quite understand. Do we really need to use all the functions of SkiaSharp? If not, why not remove the unnecessary parts?
I couldn't believe that a 2D graphics library could actually reach 10mb in size.

**Reply by hez2010** [reply]
Ideally the graphic library should be written in all managed code. But that's yet another non-trivial work needs years of human work which isn't realistic. 
If you do concern about the size of NativeAOT binary and want to get rid of the skia library, you can use static linking to link the libskia into the app directly.

**Reply by sgf** [reply]
> Ideally the graphic library should be written in all managed code. But that's yet another non-trivial work needs years of human work which isn't realistic. If you do concern about the size of NativeAOT binary and want to get rid of the skia library, you can use static linking to link the libskia into the app directly.

Nowadays, artificial intelligence and machine learning are so highly praised. Is it possible that even this simple code translation cannot be done?

**Reply by timunie** [reply]
STOP SPAMMING US!

**Reply by Cubody** [reply]
> Nowadays, artificial intelligence and machine learning are so highly praised. Is it possible that even this simple code translation cannot be done?

I am an active and old user of Chat GPT 3, 4 and 4o and I confidently declare that so far your idea is not feasible for AI =)
But maybe someday...

**Comment by peaceshi** [comment]
ANGLE.Static and SkiaSharp.Static released a new version.

ANGLE static lib now uses the official chromium branch and builds with gn.
**Binary lib size reduces 90% (1GB to 100MB).**

SkiaSharp.Static now uses the official cake generated args.gn.
**Binary lib size reduces 97% (1GB to 30MB).**

The final program size **reduces 20% (45MB to 36MB).**
**With UPX, only 11MB.**

I also create a sample project to show how to use nativeaot.

[ANGLE.Static](https://github.com/2ndlab/ANGLE.Static)
[SkiaSharp.Static](https://github.com/2ndlab/SkiaSharp.Static)

https://github.com/2ndlab/ANGLE.Static/releases/tag/v2.0.0-pre
https://github.com/2ndlab/SkiaSharp.Static/releases/tag/v2.0.0-pre

https://github.com/peaceshi/Avalonia-NativeAOT-SingleFile

https://github.com/peaceshi/Avalonia-NativeAOT-SingleFile/blob/ed4d569d55c5302671710eec38824d9b17fda5c7/Avalonia-NativeAOT-SingleFile/Avalonia-NativeAOT-SingleFile.csproj#L31-L42

**Reply by theoractice** [reply]
For this `v2.0.0-pre` version, I need to add `gdi32.lib;synchronization.lib;d3d9.lib;dxgi.lib;dxguid.lib` to the linker in VS2022.
```xml
<ItemGroup>
	<DirectPInvoke Include="libHarfBuzzSharp" />
	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\libHarfBuzzSharp.lib" />
	<DirectPInvoke Include="libSkiaSharp" />
	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\skia.lib" />
	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\SkiaSharp.lib" />
	<DirectPInvoke Include="av_libglesv2" />
	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\libANGLE_static.lib" />
	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\libGLESv2_static.lib" />
	<LinkerArg Include="gdi32.lib;synchronization.lib;d3d9.lib;dxgi.lib;dxguid.lib" />
</ItemGroup>
```

**Reply by peaceshi** [reply]
> For this `v2.0.0-pre` version, I need to add `gdi32.lib;synchronization.lib;d3d9.lib;dxgi.lib;dxguid.lib` to the linker in VS2022.
> ```xml
> <ItemGroup>
> 	<DirectPInvoke Include="libHarfBuzzSharp" />
> 	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\libHarfBuzzSharp.lib" />
> 	<DirectPInvoke Include="libSkiaSharp" />
> 	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\skia.lib" />
> 	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\SkiaSharp.lib" />
> 	<DirectPInvoke Include="av_libglesv2" />
> 	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\libANGLE_static.lib" />
> 	<NativeLibrary Include="runtimes\$(RuntimeIdentifier)\native\libGLESv2_static.lib" />
> 	<LinkerArg Include="gdi32.lib;synchronization.lib;d3d9.lib;dxgi.lib;dxguid.lib" />
> </ItemGroup>
> ```

yes，you need to add system libs depends on your cases. it is a part of reason i do not link them as a `fat(single)` lib.

**Reply by DeafMan1983** [reply]
Don't use UPX because it doesn't work during release to application. I find better if you use under Linux than it doesn't happen but some binding libraries need to load.

**Reply by peaceshi** [reply]
> Don't use UPX because it doesn't work during release to application. I find better if you use under Linux than it doesn't happen but some binding libraries need to load. 

yeah，you should make sure your program is working well，then pack it with upx.

**Reply by theoractice** [reply]
> > Don't use UPX because it doesn't work during release to application. I find better if you use under Linux than it doesn't happen but some binding libraries need to load. 
> 
> yeah，you should make sure your program is working well，then pack it with upx.

One downside of UPX is that it uses more memory to load the uncompressed executable, so it's a trade-off between memory usage and disk space.

**Reply by fvffv** [reply]
> > > Don't use UPX because it doesn't work during release to application. I find better if you use under Linux than it doesn't happen but some binding libraries need to load.
> > 
> > 
> > yeah，you should make sure your program is working well，then pack it with upx.
> 
> One downside of UPX is that it uses more memory to load the uncompressed executable, so it's a trade-off between memory usage and disk space.

Hey! I’m developing a modified version of UPX that uses the Misa77 algorithm. It can significantly reduce the additional startup time overhead introduced by UPX compression, and it also supports releasing the extra memory after the application has started.

Feel free to take a look if you’re interested:
"https://github.com/fvffv/UPXEX" (https://reference-url-citation.invalid/0)

**Reply by fvffv** [reply]
> 角度。Static 和 SkiaSharp.Static 发布了新版本。
> 
> ANGLE 静态库现在使用官方的 chromium 分支，并用 gn 构建。二**进制库大小减少了90%（1GB变为100MB）。**
> 
> SkiaSharp.Static 现在使用官方生成的蛋糕生成 Cookie 的 args.gn。 **二进制库容量减少了97%（1GB缩减为30MB）。**
> 
> 最终程序**大小减少了20%（45MB降至36MB）。** **用UPX时，只有11MB。**
> 
> 我还制作了一个示例项目，展示如何使用 Nativeaot。
> 
> [角度。静电声](https://github.com/2ndlab/ANGLE.Static) [SkiaSharp。静电声](https://github.com/2ndlab/SkiaSharp.Static)
> 
> https://github.com/2ndlab/ANGLE.Static/releases/tag/v2.0.0-pre https://github.com/2ndlab/SkiaSharp.Static/releases/tag/v2.0.0-pre
> 
> https://github.com/peaceshi/Avalonia-NativeAOT-SingleFile
> 
> https://github.com/peaceshi/Avalonia-NativeAOT-SingleFile/blob/ed4d569d55c5302671710eec38824d9b17fda5c7/Avalonia-NativeAOT-SingleFile/Avalonia-NativeAOT-SingleFile.csproj#L31-L42

Hey! I’m developing a modified version of UPX that uses the Misa77 algorithm. It can significantly reduce the additional startup time overhead introduced by UPX compression, and it also supports releasing the extra memory after the application has started.

Feel free to take a look if you’re interested:
"https://github.com/fvffv/UPXEX" (https://reference-url-citation.invalid/0)
